Automatic passenger deletion additions & fixes

Starkku opened the pull request:
- Added an option to base deletion interval on passenger cost instead of size, house filters, ability to toggle off recording kills to the transport as well as FlyingString display for refund.
- `Gunner=true` transports now correctly change their turret & weapon on deletion of a passenger.
- Previously if passenger had no explicitly set `Soylent` value, nothing would be refunded. The refund amount is now calculated in exactly same way as `Grinding=true` buildings do it.

In addition, the documentation has been updated to be more elaborate.

------------------

- Transports can erase passengers over time. Passengers are deleted in order of entering the transport, from first to last.
  - `PassengerDeletion.Rate` determines the interval in game frames that it takes to erase a single passenger.
    - If `PassengerDeletion.Rate.SizeMultiply` is set to true, this time interval is multiplied by the passenger's `Size`.
  - `PassengerDeletion.UseCostAsRate`, if set to true, changes the time interval for erasing a passenger to be based on the passenger's `Cost`. This does not factor in modifiers like `FactoryPlant`.
    - `PassengerDeletion.CostMultiplier` can be used to modify the cost-based time interval.
    - `PassengerDeletion.CostRateCap` can be used to set a cap to the cost-based time interval.
  - `PassengerDeletion.AllowedHouses` determines which houses passengers can belong to be eligible for deletion.
  - `PassengerDeletion.DontScore`, if set to true, makes it so that the deleted passengers are not counted as having been killed by the transport (no experience, not recorded towards owning house's score etc).
  - If `PassengerDeletion.Soylent` is set to true, an amount of credits is refunded to the owner of the transport. The exact amount refunded is determined by the passengers `Soylent`, or if not set, its `Cost` (this is affected by modifiers such as `FactoryPlant`).
    - `PassengerDeletion.SoylentMultiplier` is a direct multiplier applied to the refunded amount of credits.
    - `PassengerDeletion.SoylentAllowedHouses` determines which houses passengers can belong to be eligible for refunding.
    - `PassengerDeletion.DisplaySoylent` can be set to true to display the amount of credits refunded on the transport. `PassengerDeletion.DisplaySoylentToHouses` determines which houses can see this and `PassengerDeletion.DisplaySoylentOffset` can be used to adjust the display offset.
  - `PassengerDeletion.ReportSound` and `PassengerDeletion.Anim` can be used to specify a sound and animation to play when a passenger is erased, respectively.

In `rulesmd.ini`:
```ini
[SOMETECHNO]                                    ; TechnoType
PassengerDeletion.Rate=0                        ; integer, game frames
PassengerDeletion.Rate.SizeMultiply=true        ; boolean
PassengerDeletion.UseCostAsRate=false           ; boolean
PassengerDeletion.CostMultiplier=1.0            ; floating point value, percents or absolute
PassengerDeletion.CostRateCap=                  ; integer, game frames
PassengerDeletion.AllowedHouses=all             ; Affected House Enumeration (none|owner/self|allies/ally|team|enemies/enemy|all)
PassengerDeletion.DontScore=false               ; boolean
PassengerDeletion.Soylent=false                 ; boolean
PassengerDeletion.SoylentMultiplier=1.0         ; float, percents or absolute
PassengerDeletion.SoylentAllowedHouses=enemies  ; Affected House Enumeration (none|owner/self|allies/ally|team|enemies/enemy|all)
PassengerDeletion.DisplaySoylent=false          ; boolean
PassengerDeletion.DisplaySoylentToHouses=All    ; Affected House Enumeration (none|owner/self|allies/ally|team|enemies/enemy|all)
PassengerDeletion.DisplaySoylentOffset=0,0      ; X,Y, pixels relative to default
PassengerDeletion.ReportSound=                  ; Sound
PassengerDeletion.Anim=                         ; Animation
```

FS-21 commented on src/Ext/Techno/Body.cpp:
```diff
@@ -358,35 +372,51 @@ void TechnoExt::EatPassengers(TechnoClass* pThis)
+						// Check if there is money refund
+						if (pData->PassengerDeletion_Soylent)
+						{
+							int nMoneyToGive = 0;
+
+							// Refund money to the Attacker
+							nMoneyToGive = pPassenger->GetTechnoType()->GetRefund(pPassenger->Owner, true);
```
This can be merged into 1 line.

```suggestion
							// Refund money to the Attacker
							int nMoneyToGive = pPassenger->GetTechnoType()->GetRefund(pPassenger->Owner, true);
```

FS-21 commented on src/Ext/WarheadType/Detonate.cpp:
```diff
@@ -42,24 +42,12 @@ void WarheadTypeExt::ExtData::Detonate(TechnoClass* pOwner, HouseClass* pHouse,
+				auto displayCoords = this->TransactMoney_Display_AtFirer ? (pOwner ? pOwner->Location : coords) : coords;
+				FlyingStrings::AddMoneyString(this->TransactMoney, pHouse, this->TransactMoney_Display_Houses, displayCoords, this->TransactMoney_Display_Offset);
 			}
+
```
Unnecessary empty space

FS-21 reviewed (changes requested):
Only small tweaks are needed to the current code.

Starkku commented:
Fixed the issues and added `PassengerDeletion.SoylentMultiplier`.

FS-21 reviewed (dismissed):
Except one detail I'm not sure everything else looks fine.

MaxLenarc commented:
I went ahead and tested this logic. Works for the most part for me, though there were parts I could not get to work and am not sure why.
PassengerDeletion.ReportSound works as intended. PassengerDeletion.Anim works as intended. PassengerDeletion.Rate works as intended. PassengerDeletion.Soylent, PassengerDeletion.SoylentMultiplier, PassengerDeletion.SoylentFriendlies, PassengerDeletion.DisplaySoylent all work as intended.
PassengerDeletion.Rate.SizeMultiply works as intended. PassengerDeletion.DisplaySoylentOffset works as intended.
PassengerDeletion.UseCostAsRate does not work.
PassegnerDeletion.CostMultiplier does not work.
PassengerDeletion.DisplaySoylenyToHouses works as intended.

PassengerDeletion.UseCostAsRate I set to true and PassengerDeletion.UseCostAsRate I set to 100.0 and set to the Allied Landing Craft. I used the Battlefortress, Grizzly Tanks, and GIs as deletion subjects. Mainly the latter two, and found they just about always at the same time no matter what numbers I put in. Not sure If I did something wrong.

Metadorius commented:
Can you please cherry pick the fixes from new feature additions into a new PR so we include it into 0.3?

Starkku commented:
> PassengerDeletion.UseCostAsRate I set to true and PassengerDeletion.UseCostAsRate I set to 100.0 and set to the Allied Landing Craft. I used the Battlefortress, Grizzly Tanks, and GIs as deletion subjects. Mainly the latter two, and found they just about always at the same time no matter what numbers I put in. Not sure If I did something wrong.

Are you sure you did not disregard this bit from the documentation when it comes to `PassengerDeletion.UseCostAsRate`?

`If PassengerDeletion.Rate is also set to a value higher than 0, it serves as a maximum cap for the time interval.`

MaxLenarc commented:
Been o long since I tried it out initially that I don't remember. I will test it again tomorrow.

MaxLenarc commented:
You are correct it was an error on my part. I had PassengerDeletion.Rate=50 and PassengerDeletion.CostMultiplier=100.0. Will test again when able.

MaxLenarc commented:
I tested it out again and nothing worked still. Here is how the code looks. Did I mess up on anything?
PassengerDeletion.Rate=200                     ; integer, game frames
PassengerDeletion.Rate.SizeMultiply=false      ; boolean
PassengerDeletion.UseCostAsRate=true         ; boolean
PassengerDeletion.CostMultiplier=2.0         ; float, percents or absolute
PassengerDeletion.Soylent=true               ; boolean
PassengerDeletion.SoylentMultiplier=10.0       ; float, percents or absolute
PassengerDeletion.SoylentFriendlies=true     ; boolean
PassengerDeletion.DisplaySoylent=true        ; boolean
PassengerDeletion.DisplaySoylentToHouses=All  ; Affected House Enumeration (none|owner/self|allies/ally|team|enemies/enemy|all)
PassengerDeletion.DisplaySoylentOffset=100,10    ; X,Y, pixels relative to default
PassengerDeletion.ReportSound=GIAttack        ; sound
PassengerDeletion.Anim=UCFLASH                ; animation

Starkku commented:
> I tested it out again and nothing worked still. Here is how the code looks. Did I mess up on anything? 

Try with `PassengerDeletion.Rate=0` or leave omitted, or a much higher value than 200. In this case the duration of the deletion is capped to 200 frames regardless of cost, so anything that costs more than 100 credits (accounting for the multiplier) gets deleted in same time as $100 items.

MaxLenarc commented:
I already tried setting PassengerDeletion.Rate to 0. But all that did was shut off the whole unit deletion thing. Okay I got it to work. I set the PassengerDeletionRate to 20000 and then started seeing differences in time for units to get deleted. Well seems like it is good to go to me. Thanks for your help!

Starkku commented:
Fixed a bug that caused `PassengerDeletion.UseCostAsRate` to not work if `PassengerDeletion.Rate` was 0 or below and added separate `PassengerDeletion.CostRateCap` instead of reusing `PassengerDeletion.Rate` to prevent potential confusion.

Metadorius commented:
> Can you please cherry pick the fixes from new feature additions into a new PR so we include it into 0.3?

Still waiting on response on that one.

Starkku commented:
> > Can you please cherry pick the fixes from new feature additions into a new PR so we include it into 0.3?
> 
> Still waiting on response on that one.

I can but given that there's more fixes than new features, which mind you have been more or less tested at this point, I do wonder if there's any point

Metadorius commented on docs/New-or-Enhanced-Logics.md:
```diff
@@ -480,18 +480,34 @@ LimboKill.IDs=                  ; List of numeric IDs.
+PassengerDeletion.CostRateCap=                ; integer, game frames
+PassengerDeletion.Soylent=false               ; boolean
+PassengerDeletion.SoylentMultiplier=1.0       ; float, percents or absolute
+PassengerDeletion.SoylentFriendlies=false     ; boolean
```
SoylentFriendlies kinda asks for AffectedHouses instead since it's already a separate tag. Also not sure if there's a possibility to disable/enable deleting passengers per house at all?

Metadorius commented on src/Ext/Techno/Body.cpp:
```diff
@@ -120,21 +120,35 @@ void TechnoExt::ExtData::EatPassengers()
+						timerLength *= (int)(pPassenger->GetTechnoType()->Size + 0.5);
+				}
+
+				this->PassengerDeletionCountDown = timerLength;
+				this->PassengerDeletionTimer.Start(timerLength);
```
Seems kinda redundant to have both?

Metadorius commented on src/Ext/Techno/Body.cpp:
```diff
@@ -175,18 +189,35 @@ void TechnoExt::ExtData::EatPassengers()
-								// Refund money to the Attacker
-								if (pPassengerType && pPassengerType->Soylent > 0)
-									nMoneyToGive = pPassengerType->Soylent;
+								int nMoneyToGive = pPassenger->GetTechnoType()->GetRefund(pPassenger->Owner, true);
+								nMoneyToGive = (int)(nMoneyToGive * pTypeExt->PassengerDeletion_SoylentMultiplier);
```
Why not
```suggestion
								int nMoneyToGive = (int)(pPassenger->GetTechnoType()->GetRefund(pPassenger->Owner, true)
									* pTypeExt->PassengerDeletion_SoylentMultiplier);
```

Starkku commented:
Applied changes based on feedback, as well as the following:

- Set passenger deletion timer based on first passenger entered (the passenger to be deleted) rather than last
- Add `PassengerDeletion.(Soylent)AllowedHouses` to control which houses passengers can belong to be eligible for deletion and/or refunding.
  - `PassengerDeletion.SoylentFriendlies` has been deprecated as a result.
- Add `PassengerDeletion.DontScore` to control whether or not deleted passengers are counted as killed by the transport.

chaserli commented on src/Ext/TechnoType/Body.h:
```diff
@@ -46,11 +46,20 @@ class TechnoTypeExt
 		Valueable<int> PassengerDeletion_Rate;
-		NullableIdx<VocClass> PassengerDeletion_ReportSound;
 		Valueable<bool> PassengerDeletion_Rate_SizeMultiply;
+		Valueable<bool> PassengerDeletion_UseCostAsRate;
```
For the sake of readability and maintainability, I recommend a similar approach as #787 . If anyone uses this logic, there should be only a few or even one single technotype.

chaserli commented on src/Ext/Techno/Body.cpp:
```diff
@@ -159,93 +159,132 @@ void TechnoExt::ExtData::EatPassengers()
-						pPassenger = static_cast<FootClass*>(pPassenger->NextObject);
-					}
+					if (pDelType->CostRateCap.isset())
+						timerLength = std::min(timerLength, pDelType->CostRateCap.Get());
```
I've seen mixed use of `std::min` and `Math::min` in the project ( code written by different people of course). What's the difference exactly?

Starkku replied:
Probably none in the outcome, `std::min` is obviously C++ standard library implementation and `Math::min` is in YRpp

```cpp
template <typename T, typename T2>
inline auto min(T&& value, T2&& value2)
{
	if(value2 < value) {
		return static_cast<value_return_t<T>>(value2);
	}
	return value;
}
````

I would assume this is similar if not identical to `std::min` implementation, aside from the template stuff. Probably for consistency's sake should stick to one, so I guess `Math::min` which is probably used more.

chaserli commented on src/Ext/Techno/Body.cpp:
```diff
@@ -159,93 +159,132 @@ void TechnoExt::ExtData::EatPassengers()
+					auto pSource = pDelType->DontScore ? nullptr : pThis;
+					pPassenger->KillPassengers(pSource);
+					pPassenger->RegisterDestruction(pSource);
+					pPassenger->UnInit();
```
`KillPassengers` will have been called in `UnInit` too, but yeah it's idempotent so doesn't really matter.

Starkku replied:
¯\_(ツ)_/¯

chaserli reviewed (commented):
Is this finished? If not what's missing?

chaserli reviewed (approved):
Almost a year. Are you still planning to do any further modifications?
